### PR TITLE
timeout per attempt and retries

### PR DESCRIPTION
**Pull request by ShanniLi**
r @Raynos @kriskowal

### CODE REVIEW

File: node/request.js [hunk 1]
```diff
@@ -329,7 +335,7 @@ TChannelRequest.prototype.shouldRetryError = function shouldRetryError(err) {
                 return true;
 
             case 'Timeout':
-                return !!self.options.retryFlags.onTimeout;
+                return !!err.logical || !!self.options.retryFlags.onTimeout;
```
Comment (kriskowal) [inline node/request.js]: !! is superfluous before a boolean.

**Comment by kriskowal** [comment]
I believe we should look at per-attempt timeouts as a stopgap, as likely to cause harm as help since they can multiply effort in progress. Speculative trials is the endgame.

File: node/request.js (hunk 1 above)
Comment (ShanniLi) [inline node/request.js]: err.logical can be undefined

**Comment by ShanniLi** [comment]
@kriskowal we got user cases where clients time out on services occasionally. In these cases we need some kind of retry. I discussed with @Raynos, seems that this kind of retry is desired. That is reason for this PR. :)

File: node/request.js [hunk 2]
```diff
@@ -206,6 +207,10 @@ TChannelRequest.prototype.onIdentified = function onIdentified(peer) {
         opts[key] = self.options[key];
     }
     opts.timeout = self.timeout - self.elapsed;
+    if (opts.timeout > self.timeoutPerAttempt) {
```
Comment (Raynos) [inline node/request.js]: This feels wrong.

File: node/request.js (hunk 1 above)
Comment (Raynos) [inline node/request.js]: Why check for `err.logical`; just retry on all timeouts.

File: node/test/timeouts.js [hunk 3]
```diff
@@ -120,3 +120,219 @@ allocCluster.test('requests will timeout', {
...
+                timeout: 3000,
+                timeoutPerAttempt: 1000,
+                retryLimit: 2
+            })
+            .send('/timeout', 'h', 'b', done);
+    }
+
+    function done(err, res, arg2, arg3) {
+        assert.ifError(err);
+
```
Comment (Raynos) [inline node/test/timeouts.js]: the logical `req` object contains an array outReqs

You should assert on `req.outReq`; namely that there are two out reqs and that the first failed with a timeout.

File: node/request.js (hunk 1 above)
Comment (ShanniLi) [inline node/request.js]: For retry on total timeout, we don't want it to be enabled as default, right?

Comment (Raynos) [inline node/request.js]: the retry logic will check the timeout and early abort if it passed.

File: node/request.js [hunk 4]
```diff
@@ -329,7 +334,7 @@ TChannelRequest.prototype.shouldRetryError = function shouldRetryError(err) {
                 return true;
 
             case 'Timeout':
-                return !!self.options.retryFlags.onTimeout;
+                return true;
```
Comment (Raynos) [inline node/request.js]: Respect `!!self.options.retryFlags.onTimeout`; dont set this to true.

Sorry; didnt communicate properly.

File: node/request.js [hunk 5]
```diff
@@ -46,14 +46,14 @@ function TChannelRequest(channel, options) {
     if (!self.options.retryFlags) {
         self.options.retryFlags = {
             never: false,
-            onConnectionError: true,
-            onTimeout: false
```
Comment (Raynos) [inline node/request.js]: leave onTimeout: false here

Comment (Raynos) [inline node/request.js]: Actually; if `timeoutPerAttempt` is specified then set `onTimeout: true`

File: node/test/retry.js [hunk 6]
```diff
@@ -317,26 +317,12 @@ allocCluster.test('retryFlags work', {
     });
 
     series([
-
-        function defaultToNotRetryingTimeout(next) {
```
Comment (Raynos) [inline node/test/retry.js]: Why change this code ?

File: node/test/retry.js [hunk 7]
```diff
@@ -356,25 +342,7 @@ allocCluster.test('retryFlags work', {
 
                 next();
             });
-        },
-
-        function canOptOutFully(next) {
```
Comment (Raynos) [inline node/test/retry.js]: Why change this ?

File: node/v2/index.js [hunk 8]
```diff
@@ -84,11 +84,9 @@ module.exports.parseRetryFlags = function parseRetryFlags(val) {
     val = val || 'c';
     var never = val.indexOf('n') > -1;
     var onConnectionError = !never && val.indexOf('c') > -1;
-    var onTimeout = !never && val.indexOf('t') > -1;
```
Comment (Raynos) [inline node/v2/index.js]: leave the `onTimeout` flag here.

File: node/test/timeouts.js (hunk 3 above)
Comment (ShanniLi) [inline node/test/timeouts.js]: fixed

**Comment by Raynos** [comment]
lgtm. Thank you @ShanniLi

**Pull request merged**